Gauge/BarGauge: Added support for value mapping of "no data"-state to text/value

mckn opened the pull request:
**What this PR does / why we need it**:
This enables our users to map a "no data"-state and display an value/text instead of the "no data" text. I tried to prevent breaking and old behaviour by checking for mappings of null. The only case where this might happen is if null already is mapped to another value then "no data". 

**Which issue(s) this PR fixes**:
Fixes #20706

**Special notes for your reviewer**:
I wanted to move the factory method into its own module but that would require some more refactoring due to the current dependency graph. So I just settled with this solution.

mckn committed:
FieldDisplay: added support for value mapping of no data state.

 Committer: Marcus Andersson <marcus.andersson@grafana.com>

mckn commented:
Note to self: check if there is any documentation that needs to be updated regarding the value mapping functionality.

hugohaggmark commented on packages/grafana-data/src/field/fieldDisplay.test.ts:
```diff
@@ -156,4 +157,64 @@ describe('FieldDisplay', () => {
+        defaults: {},
+      },
+      theme: {} as GrafanaTheme,
+    };
```
Nitpick: maybe move this to a separate `testContext|testSetup|setup` function and reuse some of the code in both test cases?

mckn replied:
Yeah, I was thinking about this as well. But I have a couple of other ideas that I would like to check with you guys. But suuper good feedback!

I am a big fan of the builder pattern when it comes to creating test-data. Please let me know if this is something we should talk more about.

hugohaggmark replied:
Yes, please share

hugohaggmark commented on packages/grafana-data/src/field/fieldDisplay.ts:
```diff
@@ -298,3 +293,51 @@ export function getDisplayValueAlignmentFactors(values: FieldDisplay[]): Display
   return info;
 }
+
+function createNoValuesFieldDisplay(options: GetFieldDisplayValuesOptions): FieldDisplay {
```
Just out of curiosity why are these functions not tested as units but only as part of testing `getFieldDisplayValues `?

mckn replied:
Currently, because it is an internal functionality of the `fieldDisplay` module. I would have added tests for this scenario if I could have extracted it as its own module.

It is doable but require some more work/refactorings and I think I am a little bit to new to the code base to make larger changes. It is always good to get some more insight about the code base as a whole before suggesting bigger changes (if that makes sense).

hugohaggmark replied:
Thanks

dprokop replied:
> Currently, because it is an internal functionality of the fieldDisplay module. I would have added tests for this scenario if I could have extracted it as its own module.

Not sure I follow why you would need to extract it to a separate module to unit test it?

mckn replied:
Hehe, so did my reasoning make sense to you guys? Is it ok to resolve this conversion or should I take any more action on it?

torkelo replied:
no, I think this is fine

hugohaggmark commented on packages/grafana-data/src/utils/valueMappings.ts:
```diff
@@ -84,3 +84,10 @@ const getAllFormattedValueMappings = (valueMappings: ValueMapping[], value: Time
+    return false;
+  }
+  return mapping.value.toLowerCase() === 'null';
+};
```
Awesome!

hugohaggmark reviewed (changes requested):
I found some strange behaviour when testing the branch. When I change the `mode` of the BarGauge somehow the value is changed.
![Kapture 2019-12-04 at 6 53 24](https://user-images.githubusercontent.com/562238/70116374-d4bd4000-1662-11ea-8f0d-7295d0ccad33.gif)

mckn commented:
> I found some strange behaviour when testing the branch. When I change the `mode` of the BarGauge somehow the value is changed.
> ![Kapture 2019-12-04 at 6 53 24](https://user-images.githubusercontent.com/562238/70116374-d4bd4000-1662-11ea-8f0d-7295d0ccad33.gif)

Interesting, will have a look and see if my change is the causing the behaviour or if it is as expected(?). Great findings and feedback!

dprokop reviewed (commented):
Code looks good, great stuff with adding some tests! Looking forward to know what's the problem with the issue @hugohaggmark found :)

mckn committed:
FieldDisplay: fixed issue when switching between modes and display numeric was null.

torkelo commented on packages/grafana-data/src/field/fieldDisplay.ts:
```diff
@@ -298,3 +293,51 @@ export function getDisplayValueAlignmentFactors(values: FieldDisplay[]): Display
+        text: defaultDisplayText,
+      },
+    };
+  }
```
want to try to find a way to get rid of this condition, feels wrong to have to check is this mapping exists

mckn replied:
Awesome that we think alike. I have already pushed an update to simplify it a bit.

mckn commented:
> Code looks good, great stuff with adding some tests! Looking forward to know what's the problem with the issue @hugohaggmark found :)

So the reason why we got this behaviour was due to `display.numeric: null` and consumers where expecting `0` when there are no data. So I added a special test case for this scenario and changed the code a bit. Suuper good that @hugohaggmark found it!

mckn committed:
ValueMapping: introduced a private function for checking null values.

torkelo reviewed (approved):
Looks great now!

mckn committed:
FieldDisplay: refactoring of test setup to reduce duplication.

hugohaggmark reviewed (approved):
LGTM and works as expected!

mckn committed:
Docs: added info about new 'no data' value to text mapping.

mckn commented:
@oddlittlebird @marcusolsson I would be super happy if you guys could have a look at the updated documentation I have added in this PR.

marcusolsson commented on docs/sources/features/panels/singlestat.md:
```diff
@@ -23,17 +22,17 @@ The singlestat panel has a normal query editor to allow you define your exact me
 {{< docs-imagebox img="/img/docs/v45/singlestat-value-options.png" class="docs-image--no-shadow" max-width="900px" >}}
 
 1. **Stats**: The Stats field let you set the function (min, max, average, current, total, first, delta, range) that your entire query is reduced into a single value with. This reduces the entire query into a single summary value that is displayed.
```
If you'd like to improve further, I'd recommend using lazy list ordering, especially for longer lists, i.e.:

```
1. First item
1. Second item
1. Third item
```

marcusolsson commented on docs/sources/features/panels/singlestat.md:
```diff
@@ -64,7 +63,7 @@ Sparklines are a great way of seeing the historical data related to the summary
 <div class="clearfix"></div>
 
-> ***Pro-tip:*** Reduce the opacity on  fill colors for nice looking panels.
+> **_Pro-tip:_** Reduce the opacity on fill colors for nice looking panels.
```
```suggestion
> **Pro-tip:** Reduce the opacity on fill colors for nice looking panels.
```

I don't think the emphasis is needed here.

mckn replied:
ouch, mybad 🙈 I have prettier formatting enabled so it will reformat the `md` code according to the prettier settings of the project.

What is our thoughts regarding this?

marcusolsson commented on docs/sources/features/panels/singlestat.md:
```diff
@@ -85,6 +84,8 @@ Gauges gives a clear picture of how high a value is in it's context. It's a grea
 
 Value/Range to text mapping allows you to translate the value of the summary stat into explicit text. The text will respect all styling, thresholds and customization defined for the value. This can be useful to translate the number of the main Singlestat value into a context-specific human-readable word or message.
 
+If you want to replace the default "No data" text being displayed when no data is available. You can do this by adding a `value to text mapping` from `null` to your preferred custom text value.
```
```suggestion
If you want to replace the default "No data" text being displayed when no data is available, add a `value to text mapping` from `null` to your preferred custom text value.
```

The first sentence is incomplete here.

mckn replied:
awesome feedback, thanks! Will update 🙏

marcusolsson reviewed (changes requested):
Nice!

mckn committed:
Docs: improved according to feedback. Reverted prettier formatting changes.

mckn committed:
FieldDisplay: removed unused import.

gfinocchiaro commented:
Hi Guys,

it looks like that the new feature for mapping "no data" has no effect when the panel has more than one query, as you can see from the below image:

![image](https://user-images.githubusercontent.com/8842677/70801568-84a35380-1daf-11ea-957a-217212e2e786.png)

Here the mappings:
![image](https://user-images.githubusercontent.com/8842677/70801642-b61c1f00-1daf-11ea-9897-be01ff9dd7e1.png)

torkelo commented:
What are your queries returning? Is it returning a null value? 2 queries, where 1 query returns 1 series the other does not, results in 1 series, the panel (visualization) has no knowledge one query returned no data. 

So you have to modify your query so that it returns a null value when there is no data

gfinocchiaro commented:
Ok,
thank you.
